pkg/host/internal/infiniband/ib_guid_config.go

```diff
@@ -0,0 +1,131 @@
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/host/internal/lib/netlink"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/host/types"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/utils"
+)
+
+type ibPfGUIDJSONConfig struct {
+	PciAddress string   `json:"pciAddress,omitempty"`
+	PfGUID     string   `json:"pfGuid,omitempty"`
+	GUIDs      []string `json:"guids,omitempty"`
+	RangeStart string   `json:"rangeStart,omitempty"`
+	RangeEnd   string   `json:"rangeEnd,omitempty"`
+}
+
+type ibPfGUIDConfig struct {
+	GUIDs      []net.HardwareAddr
+	RangeStart utils.GUID
+	RangeEnd   utils.GUID
+}
+
+func getIbGUIDConfig(configPath string, netlinkLib netlink.NetlinkLib, networkHelper types.NetworkInterface) (map[string]ibPfGUIDConfig, error) {
+	links, err := netlinkLib.LinkList()
+	if err != nil {
+		return nil, err
+	}
+
+	rawConfigs, err := readJSONConfig(configPath)
+	if err != nil {
+		return nil, err
+	}
+
+	resultConfigs := map[string]ibPfGUIDConfig{}
+
+	// Parse JSON config into an internal struct
+	for _, rawConfig := range rawConfigs {
+		pciAddress, err := getPfPciAddressFromRawConfig(rawConfig, links, networkHelper)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract pci address from ib guid config: %w", err)
+		}
+
+		// GUID list and GUID range can't be set at the same time
+		if len(rawConfig.GUIDs) != 0 && (rawConfig.RangeStart != "" || rawConfig.RangeEnd != "") {
+			return nil, fmt.Errorf("either guid list or guid range should be provided, got both")
+		} else if (rawConfig.RangeStart != "" && rawConfig.RangeEnd == "") || (rawConfig.RangeStart == "" && rawConfig.RangeEnd != "") {
+			return nil, fmt.Errorf("both guid rangeStart and rangeEnd should be provided, got one")
+		} else if len(rawConfig.GUIDs) != 0 {
+			var guids []net.HardwareAddr
+
+			for _, guidStr := range rawConfig.GUIDs {
+				guid, err := net.ParseMAC(guidStr)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse ib guid %s: %w", guidStr, err)
+				}
+				guids = append(guids, guid)
+			}
+			resultConfigs[pciAddress] = ibPfGUIDConfig{
+				GUIDs: guids,
+			}
+		} else if rawConfig.RangeStart != "" && rawConfig.RangeEnd != "" { // use guid range
+			rangeStart, err := utils.ParseGUID(rawConfig.RangeStart)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse ib guid range start: %w", err)
+			}
+
+			rangeEnd, err := utils.ParseGUID(rawConfig.RangeEnd)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse ib guid range end: %w", err)
+			}
+
+			if rangeEnd <= rangeStart {
+				return nil, fmt.Errorf("range end cannot be less then or equal to range start")
+			}
+
+			resultConfigs[pciAddress] = ibPfGUIDConfig{
+				RangeStart: rangeStart,
+				RangeEnd:   rangeEnd,
+			}
+		} else {
+			return nil, fmt.Errorf("either guid list or guid range should be provided, got none")
+		}
+	}
+
+	return resultConfigs, nil
+}
```

Comment:
```suggestion
func getIbGUIDConfig(configPath string, netlinkLib netlink.NetlinkLib, networkHelper types.NetworkInterface) (map[string]ibPfGUIDConfig, error) {
	links, err := netlinkLib.LinkList()
	if err != nil {
		return nil, err
	}

	rawConfigs, err := readJSONConfig(configPath)
	if err != nil {
		return nil, err
	}

	resultConfigs := map[string]ibPfGUIDConfig{}

	// Parse JSON config into an internal struct
	for _, rawConfig := range rawConfigs {
		pciAddress, err := getPfPciAddressFromRawConfig(rawConfig, links, networkHelper)
		if err != nil {
			return nil, fmt.Errorf("failed to extract pci address from ib guid config: %w", err)
		}

		// GUID list and GUID range can't be set at the same time
		if len(rawConfig.GUIDs) != 0 && (rawConfig.RangeStart != "" || rawConfig.RangeEnd != "") {
			return nil, fmt.Errorf("either guid list or guid range should be provided, got both")
		}
		if (rawConfig.RangeStart != "" && rawConfig.RangeEnd == "") || (rawConfig.RangeStart == "" && rawConfig.RangeEnd != "") {
			return nil, fmt.Errorf("both guid rangeStart and rangeEnd should be provided, got one")
		}
		if len(rawConfig.GUIDs) != 0 {
			var guids []net.HardwareAddr

			for _, guidStr := range rawConfig.GUIDs {
				guid, err := net.ParseMAC(guidStr)
				if err != nil {
					return nil, fmt.Errorf("failed to parse ib guid %s: %w", guidStr, err)
				}
				guids = append(guids, guid)
			}
			resultConfigs[pciAddress] = ibPfGUIDConfig{
				GUIDs: guids,
			}
			continue
		}
		
		rangeStart, err := utils.ParseGUID(rawConfig.RangeStart)
		if err != nil {
			return nil, fmt.Errorf("failed to parse ib guid range start: %w", err)
		}

		rangeEnd, err := utils.ParseGUID(rawConfig.RangeEnd)
		if err != nil {
			return nil, fmt.Errorf("failed to parse ib guid range end: %w", err)
		}

		if rangeEnd <= rangeStart {
			return nil, fmt.Errorf("range end cannot be less then or equal to range start")
		}

		resultConfigs[pciAddress] = ibPfGUIDConfig{
			RangeStart: rangeStart,
			RangeEnd:   rangeEnd,
		}
	}

	return resultConfigs, nil
}
```

Reply:
Thx for the suggestion, changed the function (also added a missing case for when neither guid list nor guid range are provided)